hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/realtime/HoodieMergeOnReadSnapshotReader.java

```diff
@@ -0,0 +1,192 @@
+    this.logRecordScanner = getMergedLogRecordScanner();
+    LOG.debug("Time taken to scan log records: {}", timer.endTimer());
+    this.baseFileReader = getBaseFileReader(new Path(baseFilePath), jobConf);
+    this.logRecordsByKey = logRecordScanner.getRecords();
+    Set<String> logRecordKeys = new HashSet<>(this.logRecordsByKey.keySet());
+    this.mergedRecordsByKey = new ExternalSpillableMap<>(
+        getMaxCompactionMemoryInBytes(jobConf),
+        jobConf.get(SPILLABLE_MAP_BASE_PATH_PROP, DEFAULT_SPILLABLE_MAP_BASE_PATH),
+        new DefaultSizeEstimator(),
+        new HoodieRecordSizeEstimator(readerSchema),
+        jobConf.getEnum(SPILLABLE_DISK_MAP_TYPE.key(), SPILLABLE_DISK_MAP_TYPE.defaultValue()),
+        jobConf.getBoolean(DISK_MAP_BITCASK_COMPRESSION_ENABLED.key(), DISK_MAP_BITCASK_COMPRESSION_ENABLED.defaultValue()));
+    try (ClosableIterator<String> baseFileIterator = baseFileReader.getRecordKeyIterator()) {
+      timer.startTimer();
+      while (baseFileIterator.hasNext()) {
+        String key = baseFileIterator.next();
+        if (logRecordKeys.contains(key)) {
+          logRecordKeys.remove(key);
+          Option<HoodieAvroIndexedRecord> mergedRecord = buildGenericRecordWithCustomPayload(logRecordsByKey.get(key));
+          if (mergedRecord.isPresent()) {
+            HoodieRecord hoodieRecord = mergedRecord.get().copy();
+            mergedRecordsByKey.put(key, hoodieRecord);
+          }
+        }
+      }
+    }
+    LOG.debug("Time taken to merge base file and log file records: {}", timer.endTimer());
+    this.recordsIterator = mergedRecordsByKey.values().iterator();
+  }
+
+  @Override
+  public boolean hasNext() {
+    return recordsIterator.hasNext();
+  }
+
+  @Override
+  public HoodieRecord next() {
+    return recordsIterator.next();
+  }
+
+  public Map<String, HoodieRecord> getRecordsByKey() {
+    return mergedRecordsByKey;
+  }
+
+  public Iterator<HoodieRecord> getRecordsIterator() {
+    return recordsIterator;
+  }
+
+  public Map<String, HoodieRecord> getLogRecordsByKey() {
+    return logRecordsByKey;
+  }
+
+  private static HoodieRealtimeFileSplit getRealtimeSplit(String tableBasePath, String baseFilePath,
+                                                          List<HoodieLogFile> logFilePaths,
+                                                          String latestInstantTime,
+                                                          long start, long length, String[] hosts) {
+    HoodieRealtimePath realtimePath = new HoodieRealtimePath(
+        new Path(baseFilePath).getParent(),
+        baseFilePath,
+        tableBasePath,
+        logFilePaths,
+        latestInstantTime,
+        false, // TODO: Fix this to support incremental queries
+        Option.empty());
+    return HoodieInputFormatUtils.createRealtimeFileSplit(realtimePath, start, length, hosts);
+  }
+
+  private HoodieMergedLogRecordScanner getMergedLogRecordScanner() {
```

Comment:
Looks like the same scanner is also used by `RealtimeCompactedRecordReader`.  Shall we make this a util method?

Reply:
there are some changes wrt the arguments in builder methods, so kept it separate. Keeping separate is also more flexible for now. If I don't see too much differentiation i'll extract to a separate method.

Reply:
Sg